src/EigenvalueSolverTypes.f90

```diff
@@ -253,6 +253,52 @@ SUBROUTINE evsolver_sub_absintfc(solver)
   CONTAINS
 !
 !-------------------------------------------------------------------------------
+!> @brief Create an Eigenvalue solver object compatible with the provided matrix
+!>
+!> @param solver a pointer to the base eigenvalue solver type. Should be NULL
+!> @param MPIEnv the MPI environment to be used to initialize the solver
+!> @param M the matrix to use to determine which type of solver to create
+!> @param params A parameter list to use in allocating the solver
+!> 
+!> This is a simple abstract factory routine for the base EigenvalueSolverType.
+!> It uses the type of the passed MatrixType object to determine which type of
+!> eigenvalue solver to allocate, then initialized the solver with the passed
+!> parameters. The purpose of such a routine is to hide from the client code the
+!> actual type of the matrix/solver pair, and the preprocessor guards necessary
+!> to safely use the TPLs.
+    SUBROUTINE EigenvalueSolverFactory(solver,MPIEnv,M,params)
+      CHARACTER(LEN=*),PARAMETER :: myName="EigenvalueSolverFactory"
+      CLASS(EigenvalueSolverType_Base),INTENT(INOUT),POINTER :: solver
+      TYPE(MPI_EnvType),INTENT(IN),TARGET :: MPIEnv
+      CLASS(MatrixType),INTENT(IN),TARGET :: M
+      TYPE(ParamType),INTENT(IN) :: params
+      
+      IF(ASSOCIATED(solver)) THEN
+        CALL eEigenvalueSolverType%raiseError(modName//"::"//myName//" - "// &
+          "Solver pointer is already associated")
+        RETURN
+      ENDIF
+
+      SELECTTYPE(M)
+#ifdef FUTILITY_HAVE_PETSC
```

Comment:
Ifdefing out the types themselves cuts down on the number of ifdefs needed throughout, lets us get rid of all of the errors that are getting thrown way too late in the game anyways, and with the addition of an abstract factory, the types shouldn't be needed in pretty much any client code to begin with. If there is a need for client code to interact with the TPL-backed types directly, they should use their own ifdef guards.

Make sense/seem reasonable?